File: compiler/bench/totient.fs
\ expected: 30393485
\ Euler's totient function - count coprime numbers

: gcd ( a b -- gcd )
  begin dup while
    tuck mod
  repeat drop ;

: totient ( n -- phi )
  dup 1 do
    dup i gcd 1 = if swap 1+ swap then
  loop drop ;

: main
  0
  10000 1 do
    i totient +
  loop
  . cr ;

File: compiler/bench/totient.c
// expected: 30393485
// Euler's totient function - count coprime numbers
#include <stdio.h>

int gcd(int a, int b) {
    while (b) {
        int t = b;
        b = a % b;
        a = t;
    }
    return a;
}

int totient(int n) {
    int count = 0;
    for (int i = 1; i < n; i++) {
        if (gcd(n, i) == 1) count++;
    }
    return count;
}

int main(void) {
    long sum = 0;
    for (int i = 1; i < 10000; i++) {
        sum += totient(i);
    }
    printf("%ld\n", sum);
    return 0;
}
